airflow-core/src/airflow/api_fastapi/core_api/services/public/task_instances.py

```diff
@@ -0,0 +1,249 @@
+    """Service for handling bulk operations on task instances."""
+
+    def __init__(
+        self,
+        session: Session,
+        request: BulkBody[BulkTaskInstanceBody],
+        dag_id: str,
+        dag_run_id: str,
+        dag_bag: DagBagDep,
+        user: GetUserDep,
+    ):
+        super().__init__(session, request)
+        self.dag_id = dag_id
+        self.dag_run_id = dag_run_id
+        self.dag_bag = dag_bag
+        self.user = user
+
+    def categorize_task_instances(self, task_ids: set) -> tuple[dict, set, set]:
+        """
+        Categorize the given task_ids into matched_task_ids and not_found_task_ids based on existing task_ids.
+
+        Existed task instances are returned as a dict of {task_id : TaskInstance}.
+
+        :param task_ids: set of task_ids
+        :return: tuple of dict of existed task instances, set of matched task_ids, set of not found task_ids
+        """
+        existed_task_instances = self.session.execute(select(TI).filter(TI.task_id.in_(task_ids))).scalars()
+        existed_task_instances_dict = {
+            task_instance.task_id: task_instance for task_instance in existed_task_instances
+        }
+        matched_task_ids = set(existed_task_instances_dict.keys())
+        not_found_task_ids = task_ids - matched_task_ids
+        return existed_task_instances_dict, matched_task_ids, not_found_task_ids
+
+    def handle_bulk_create(
+        self, action: BulkCreateAction[BulkTaskInstanceBody], results: BulkActionResponse
+    ) -> None:
+        results.errors.append(
+            {
+                "error": "Task instances bulk create is not supported",
+                "status_code": status.HTTP_405_METHOD_NOT_ALLOWED,
+            }
+        )
+
+    def handle_bulk_update(
+        self, action: BulkUpdateAction[BulkTaskInstanceBody], results: BulkActionResponse
+    ) -> None:
+        """Bulk Update Task Instances."""
+        to_update_task_ids = {task_instance.task_id for task_instance in action.entities}
+        _, matched_task_ids, not_found_task_ids = self.categorize_task_instances(to_update_task_ids)
+
+        try:
+            if action.action_on_non_existence == BulkActionNotOnExistence.FAIL and not_found_task_ids:
+                raise HTTPException(
+                    status_code=status.HTTP_404_NOT_FOUND,
+                    detail=f"The task instances with these task_ids: {not_found_task_ids} were not found.",
+                )
+            if action.action_on_non_existence == BulkActionNotOnExistence.SKIP:
+                update_task_ids = matched_task_ids
+            else:
+                update_task_ids = to_update_task_ids
+
+            for task_instance in action.entities:
+                if task_instance.map_index is None:
+                    task_instance.map_index = -1
+
+                if task_instance.task_id in update_task_ids:
+                    dag, ti, data = _patch_ti_validate_request(
```

Comment:
```suggestion
                    dag, tis, data = _patch_ti_validate_request(
```
This is a list at this stage. (tis)